sdks/java/core/src/main/java/org/apache/beam/sdk/io/TextSource.java

```diff
@@ -487,4 +458,146 @@ private boolean readCustomLine() throws IOException {
+                + length
+                + ", limit: "
+                + count);
+      }
+
+      return new String(buf, offset, length, charset);
+    }
+  }
+
+  /**
+   * A state machine to match the delimiter in a byte stream.
+   *
+   * <pre>{@code
+   * DelimiterFinder finder = new DelimiterFinder([65, 65, 66]); // "AAB"
+   * finder.feed(65); // false. "A"
+   * finder.feed(66); // false. "AB"
+   * finder.feed(65); // false. "ABA"
+   * finder.feed(65); // false. "ABAA"
+   * finder.feed(66); // true.  "ABAAB"
+   * finder.feed(65); // false. "ABAABA"
+   *
+   * }</pre>
+   */
+  static class DelimiterFinder {
+    private final byte[] delimiter;
+    private final int[] subsequences;
+    private final int[][] trans;
+
+    int position;
+
+    public DelimiterFinder(byte[] delimiter) {
+      this.delimiter = delimiter;
+      subsequences = new int[delimiter.length];
+      trans = new int[delimiter.length][256];
+      compile();
+    }
+
+    public boolean feed(byte b) {
+      position = trans[position][byteToIndex(b)];
+      if (position == delimiter.length) {
+        position = 0;
+        return true;
+      } else {
+        return false;
+      }
+    }
+
+    public void reset() {
+      position = 0;
+    }
+
+    public String describe() {
+      StringBuilder sb = new StringBuilder();
+      sb.append("delimiter:\n").append(Bytes.asList(delimiter)).append('\n');
+      sb.append("subsequences and trans:\n");
+      for (int i = 0; i < delimiter.length; ++i) {
+        sb.append(i).append(" (").append(subsequences[i]).append("): ");
+        for (int b = Byte.MIN_VALUE; b <= Byte.MAX_VALUE; ++b) {
+          int tran = trans[i][byteToIndex((byte) b)];
+          if (tran > 0) {
+            sb.append(b).append(" -> ").append(tran).append(", ");
+          }
+        }
+        sb.append("\n");
+      }
+      sb.append("current position: ").append(position).append("\n");
+      return sb.toString();
+    }
+
+    private void compile() {
+      // e.g. "AABAAC":
+      // 0 -> 0, ""      -> ""
+      // 1 -> 0: "A"     -> ""
+      // 2 -> 1: "AA"    -> "A"
+      // 3 -> 0: "AAB"   -> "B"
+      // 4 -> 1: "AABA"  -> "A"
+      // 5 -> 2: "AABAA" -> "AA"
+
+      for (int i = 2; i < delimiter.length; i++) {
+        if (delimiter[i - 1] == delimiter[subsequences[i - 1]]) {
+          subsequences[i] = subsequences[i - 1] + 1;
+        } else if (delimiter[i - 1] == delimiter[0]) {
+          subsequences[i] = 1;
+        } else {
+          subsequences[i] = 0;
+        }
+      }
+      // e.g. "AABAAC":
+      // index (subsequence): trans
+      // 0 (0): "A" -> 1
+      // 1 (0): "A" -> 2
+      // 2 (1): "A" -> 2, "B" -> 3
+      // 3 (0): "A" -> 4
+      // 4 (1): "A" -> 5
+      // 5 (2): "A" -> 2, "B" -> 3, "C" -> 6
+      trans[0][byteToIndex(delimiter[0])] = 1;
+
+      for (int i = 1; i < delimiter.length; i++) {
+        for (int b = Byte.MIN_VALUE; b <= Byte.MAX_VALUE; b++) {
```

Comment:
would https://en.wikipedia.org/wiki/Knuth%E2%80%93Morris%E2%80%93Pratt_algorithm be preferrable? Nice to have a documented algorithm and the table is more compact. We can also remove the byteToIndex in that case.

Reply:
Thanks. I changed it to Knuth–Morris–Pratt algorithm.